### PR TITLE
Add AggressiveInlining to Double.CompareTo() and Single.CompareTo()

### PR DESCRIPTION
**Pull request by rickbrew**
As per discussion in #56493

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=56501) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=56501) before we can accept your contribution.<br/><br/>:x: rickbrew [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=56501)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=56501) it.</sub>

File: src/libraries/System.Private.CoreLib/src/System/Double.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Double.cs:191]: How common are NaNs in the scenarios involving CompareTo?  If we're going to increase call sites for this (including generic specialization), I'm wondering if we should push out the extra asm for handling NaNs into a separate method, e.g.
```C#
    public int CompareTo(double value)
    {
        if (m_value < value) return -1;
        if (m_value > value) return 1;
        if (m_value == value) return 0;
        return CompareToWithAtLeastOneNan(value);
    }
    
    private int CompareToWithAtLeastOneNan(double value)
    {
        Debug.Assert(double.IsNaN(m_value) || double.IsNaN(value));
        if (double.IsNaN(m_value))
            return double.IsNaN(value) ? 0 : -1;
        else
            return 1;
    }
```

**Comment by stephentoub** [comment]
cc: @tannergooding, @GrabYourPitchforks

File: src/libraries/System.Private.CoreLib/src/System/Double.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:191]: @stephentoub, it depends on scenario. `NaN` is commonly used as a marker value (including in scenarios such as WPF or ML.NET).

This is really a question of inlinability and so provided the JIT can inline the base method, particularly where constants are inlined, I think we'll be fine. In particular, the JIT kind-of "gives up" on certain optimizations when it knows it has a `NaN` and it could have better awareness of it. Likewise, this entire method could be intrinsified to effectively a single `ucomiss` instruction + the relevant branches (or even `cmovcc`) which would likewise greatly reduce the overall cost.

Comment (rickbrew) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:191]: For my code, `NaN` generally indicates an error -- usually a failure to deal with the not-so-subtleties of floating point precision, or not checking for `NaN` earlier in a series of computations.

For the ML / marker stuff, does it also use `CompareTo`? If sorting a list of values that contain `NaN`s, they should all slide to the end of the sorting order (if I'm reading the code correctly).

I do really like the idea of reimplementing this using intrinsics to avoid the branching altogether and also improve its inlining eligibility. Is that possible with the current state of .NET intrinsics?

**Comment by GrabYourPitchforks** [comment]
I agree with Tanner that special-casing this in the JIT to emit specialized instructions would result in the best codegen. I don't know how difficult that would be in practice, as you don't want to end up in a situation where we're really effective at setting _eax_ to the appropriate value, but where we still incur the series of jumps after the fact. Ideally you'd want the JIT to be able to generate jumps directly to the applicable targets.

**Comment by stephentoub** [comment]
Sounds like the general consensus is this should be merged as-is but there should be an issue for .NET 7 tracking making CompareTo an intrinsic?

**Comment by rickbrew** [comment]
> Sounds like the general consensus is this should be merged as-is but there should be an issue for .NET 7 tracking making CompareTo an intrinsic?

Do you want to changes you suggested above with splitting `NaN` handling into its own method?

**Comment by stephentoub** [comment]
I'll defer to Tanner/Levi's judgement on that.

**Comment by rickbrew** [comment]
If the method is split, as you suggested, I believe performance would remain about the same for datasets with large amounts (>99%, although I'm guessing/estimating here) of `NaN` values, although codegen size would increase due to inlining the first part of the method. But, sorting a list of all `NaN`s seems like an uncommon/pathological corner case that doesn't need to be considered w.r.t. optimization? @tannergooding might know better though.

**Comment by tannergooding** [comment]
I agree that even when you have `NaN` it's likely to be more of a corner case than not. However, I would like to see more numbers/data of how this impacts sparse data sets or other scenarios first.

Likewise, I think that overall the best thing is to get the JIT to treat this intrinsicly. Today we have:
```csharp
if (m_value < value) return -1;
        if (m_value > value) return 1;
        if (m_value == value) return 0;

        // At least one of the values is NaN.
        if (double.IsNaN(m_value))
            return double.IsNaN(value) ? 0 : -1;
        else
            return 1;
```

which generates assembly that is effectively:
```asm
    L0000: vzeroupper
    L0003: vmovsd xmm0, [rcx]
    L0007: vucomisd xmm1, xmm0
    L000b: ja short L0029
    L000d: vucomisd xmm0, xmm1
    L0011: ja short L0032
    L0013: vucomisd xmm0, xmm1
    L0017: jp short L001b
    L0019: je short L002f
    L001b: vucomisd xmm0, xmm0
    L001f: jp short L0023
    L0021: je short L0032
    L0023: vucomisd xmm1, xmm1
    L0027: jp short L002f
    L0029: mov eax, 0xffffffff
    L002e: ret
    L002f: xor eax, eax
    L0031: ret
    L0032: mov eax, 1
    L0037: ret
```

While in practice, we only need no more than 3 `ucomisd` (1 if neither input is `NaN`) and then a brief set of jumps and we could generate something more like:
```asm
entry:
  xor      eax,  eax     ; Clear result to zero
  vmovsd   xmm0, [rcx]   ; Load "this" into xmm0
  vucomisd xmm0, xmm1    ; Compare "this" to "value"
  jp       nan           ; Parity flag is set, one input is NaN (unpredicted)
  jc       less_than     ; Carry flag is set, "this" is less than "value"
  jnz      greater_than  ; Zero flag is not set, "this" is greater than "value"
  ret                    ; Return value is already zero (results are equal)
greater_than:
  inc      eax           ; Return value should be one
  ret
less_than:
  dec      eax           ; Return value should be negative one
  ret
nan:
  vucomisd xmm0, xmm0    ; Compare "this" to "this"
  jnp      greater_than  ; Parity flag is not set, "this" is not NaN; "value" is a NaN
  vucomisd xmm0, xmm0    ; Compare "value" to "value"
  jnp      less_than     ; Parity flag is not set, "this" is NaN; "value" is not a NaN
  ret                    ; Return value is already zero (results are equal)
```

This would result in approx. 36 bytes of assembly which is down from 56 bytes. It has significantly fewer comparisons which saves us a number of cycles (~3-7 cycles per `ucomis` on modern CPUs) and many fewer branches in the "tight" window which should improve the branch predictor throughput.

Intrinsicly recognizing `CompareTo` might also be better for scenarios where users are relying on `CompareTo` for comparison, as we could then correctly optimize something like `if (value.CompareTo(other) < 0)` to be just `value < other` and not generate the rest of the code. Although I'm not sure how useful that would be in practice.

File: src/libraries/System.Private.CoreLib/src/System/Double.cs
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:191]: @rickbrew the cutoff for .NET 6.0 is Aug 17th. If the code written like above in two methods would be a strict improvement over today, my suggestion would be to do that.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:191]: Oh, I see there's discussion of that below. But, the date applies.

**Comment by danmoseley** [comment]
@rickbrew it sounds like the next action here is to do some quick perf tests including lots of NaN's.

@tannergooding do we not have any perf tests for comparing numbers? Maybe I'm missing them..
https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.Runtime/Perf.Double.cs

**Comment by danmoseley** [comment]
@rickbrew do you expect to be able to gather this data in time to possibly get this into 6.0?

**Comment by rickbrew** [comment]
@danmoseley Cutoff for .NET 6 is 8/17? No, that won't be possible for me. 

I also wasn't sure if that was up to me or @tannergooding, since #56493 was assigned to him. Chalk it up to being a first-timer in the dotnet/runtime repo.

I can still take a look in a bit, I've been very busy finishing, preparing, and stabilizing Paint.NET 4.3 (finally ported to .NET 5)

**Comment by danmoseley** [comment]
Congratulations on the port! Do you notice any differences compared to running it on .NET Framework?

**Comment by rickbrew** [comment]
Oh yes, lots. It's faster, maybe by about 15% across the board before doing any further optimizations that weren't possible with Framework.. Except for startup performance which has about a 30% penalty due to using a lot of C++/CLI code, which can't be crossgen'd. I interop heavily with the likes of D2D, WIC, et. al. That will slowly evaporate/improve as I port all of that over to C# (@tannergooding 's TerraFX will be put to good use).

SCD is fantastic for simplifying and alleviating a lot of various install issues. Installation and updating is much faster because I crossgen on my build machine instead of NGEN on the user's system. I've also been able to sink a lot of time into SIMD optimizations, both to add newly optimized code paths and to port many others from native C code that I was p/invoking. ARM64 support is great, really glad that landed in 5.0.9 for WinForms and WPF. Being able to load plugins into isolated `AssemblyLoadContext`s is a good thing and solves a lot of issues.

Plugin compatibility remains a struggle, although most work fine as-is w/o recompilation (even old .NET 2.0 or even 1.1 DLLs). Some framework DLLs have been removed (`System.Windows.Forms.DataVisualization`, WCF), some classes have been removed (`Menu` and `ContextMenu`), and a few things have bugs in them (`XmlSerializer` dies when used in a collectible `AssemblyLoadContext`, but that's fixed in .NET 6 and I have a workaround). I'm steadily working through as many of those as possible with various workarounds, fixes, or even hot patching via Mono.Cecil. Some plugin authors will be putting out new updates, while others have wandered elsewhere in life and so their plugins will need app-side fixes/workarounds/shims.

Download size is also a struggle. SCD grows my installer from ~12MB to >50MB, so hopefully my hosting provider doesn't get stressed out. I'm also actively pursuing ways to trim the size through various creative methods (e.g. https://github.com/dotnet/runtime/issues/56699).

Having access to all the new stuff in the runtime and framework, more libraries on nuget, and the massively improved JIT/codegen is really great. It finally feels like I'm working _with_ the runtime now, instead of working around it.

**Comment by danmoseley** [comment]
That's a great result so far thank you for sharing cc @richlander although I'm guessing he's aware of your progress.

**Comment by danmoseley** [comment]
@tannergooding i will be out next week. Do you think it would be reasonable to take this change Monday with the evidence we have? It seems we have daily good confidence it will on balance be a benefit. Your call.

**Comment by danmoseley** [comment]
One other question @rickbrew as it wasn't clear - have you experimented with trimming before crossgen? We did a lot of work to make the shared framework safe to trim. WOF/Winforms are not as you likely know. Cc @eerhardt

**Comment by rickbrew** [comment]
@danmoseley I haven't tried trimming just yet, partly because I couldn't get ILLink's command-line parameters to work. But also, I suspect PDN's trimmability is low because of the plugin situation. They need access to both PDN and runtime/framework DLLs. I may look into it more later, after 4.3's release, maybe after .NET 6's release. I see value in being able to trim libraries like `Newtonsoft.Json.dll` and will definitely need it when I start using `TerraFX.Interop.dll` (8MB!), neither of which plugins would be using (and they can have their own private copy now anyway, thanks to `AseemblyLoadContext`).

Also, the biggest bang-for-buck action I can see right now is trimming out R2R data from large framework DLLs that are not used, or not used enough, at least not on the startup path (e.g. PresentationFramework.dll is 17MB, but only 6MB w/o R2R data, see also #56699). An `uncrossgen` utility would be really helpful.

**Comment by rickbrew** [comment]
I decided to start running benchmarks and gathering the data anyway, as I could use a break from the PDN stuff.

Anyway the first result I'm seeing is that @stephentoub 's "split inlined" idea is actually faster than the fully inlined version, at least on NaN-less data. Still fleshing out the benchmark to get data on various mixtures of NaNs in the data, but I wanted to share that first before anything got merged.

**Comment by rickbrew** [comment]
I've created some benchmark code in this repository: https://github.com/rickbrew/CompareInliningBenchmarks

VS 2022 17.0.0 Preview 3.0
Ryzen 5950X, PBO enabled, 64GB DDR4-3600 RAM
Because PBO (precision boost overdrive, aka auto-overclocking) is temperature sensitive, I'm running the CPU fans at max to help it run at a consistent clock speed. Otherwise it might run at a slower speed later in the benchmark run. There's also a portable AC unit in the room.

The benchmark consists of creating an array with several million elements, sized so as to fit just within the 32MB L3 cache available to each chiplet on the CPU. The array is filled with values 0 through N-1 (ints cast to float/double). The `NaNPercentage` determines how many NaNs are evenly spread throughout the array. Then the array is shuffled using Fisher-Yates. The random seed is the same every time, for both floats and doubles.

Here are the results for `double`s with 4M elements in the array:

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1165 (21H1/May2021Update)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.100-preview.7.21379.14
  [Host]     : .NET 6.0.0 (6.0.21.37719), X64 RyuJIT
  Job-JXESFV : .NET 6.0.0 (6.0.21.37719), X64 RyuJIT
```

|           Method | NaNPercentage |     Mean |   Error |   StdDev |   Median | Ratio |
|----------------- |-------------- |---------:|--------:|---------:|---------:|------:|
|     SortStandard |             0 | 296.9 ms | 2.36 ms |  2.20 ms | 297.6 ms |  1.00 |
| SortFullyInlined |             0 | 192.4 ms | 3.58 ms |  3.35 ms | 191.4 ms |  0.65 |
| SortSplitInlined |             0 | 197.1 ms | 2.74 ms |  2.57 ms | 197.2 ms |  0.66 |
|                  |               |          |         |          |          |       |
|     SortStandard |            10 | 299.1 ms | 2.45 ms |  2.29 ms | 299.6 ms |  1.00 |
| SortFullyInlined |            10 | 193.1 ms | 3.20 ms |  2.99 ms | 192.6 ms |  0.65 |
| SortSplitInlined |            10 | 197.4 ms | 3.03 ms |  2.84 ms | 196.3 ms |  0.66 |
|                  |               |          |         |          |          |       |
|     SortStandard |            20 | 295.4 ms | 4.62 ms |  4.33 ms | 295.2 ms |  1.00 |
| SortFullyInlined |            20 | 187.6 ms | 3.65 ms |  4.21 ms | 186.8 ms |  0.64 |
| SortSplitInlined |            20 | 192.1 ms | 2.60 ms |  2.43 ms | 192.2 ms |  0.65 |
|                  |               |          |         |          |          |       |
|     SortStandard |            30 | 293.2 ms | 5.50 ms |  5.40 ms | 291.7 ms |  1.00 |
| SortFullyInlined |            30 | 177.0 ms | 3.54 ms |  7.14 ms | 173.9 ms |  0.63 |
| SortSplitInlined |            30 | 192.1 ms | 3.82 ms |  6.69 ms | 189.9 ms |  0.68 |
|                  |               |          |         |          |          |       |
|     SortStandard |            40 | 275.6 ms | 2.54 ms |  2.26 ms | 275.7 ms |  1.00 |
| SortFullyInlined |            40 | 166.0 ms | 3.29 ms |  5.94 ms | 163.9 ms |  0.63 |
| SortSplitInlined |            40 | 184.7 ms | 3.62 ms |  4.17 ms | 184.7 ms |  0.68 |
|                  |               |          |         |          |          |       |
|     SortStandard |            50 | 271.7 ms | 5.40 ms | 10.27 ms | 268.9 ms |  1.00 |
| SortFullyInlined |            50 | 158.7 ms | 3.17 ms |  6.69 ms | 156.7 ms |  0.59 |
| SortSplitInlined |            50 | 180.1 ms | 3.60 ms |  7.18 ms | 177.6 ms |  0.67 |

And for floats with 8M elements in the array. Same amount of memory and bandwidth, but higher compute cost:
|           Method | NaNPercentage |     Mean |   Error |  StdDev | Ratio |
|----------------- |-------------- |---------:|--------:|--------:|------:|
|     SortStandard |             0 | 867.2 ms | 1.69 ms | 1.49 ms |  1.00 |
| SortFullyInlined |             0 | 640.1 ms | 2.66 ms | 2.22 ms |  0.74 |
| SortSplitInlined |             0 | 670.8 ms | 3.08 ms | 2.88 ms |  0.77 |
|                  |               |          |         |         |       |
|     SortStandard |            10 | 827.5 ms | 5.48 ms | 5.13 ms |  1.00 |
| SortFullyInlined |            10 | 642.7 ms | 3.07 ms | 2.87 ms |  0.78 |
| SortSplitInlined |            10 | 661.3 ms | 3.64 ms | 3.40 ms |  0.80 |
|                  |               |          |         |         |       |
|     SortStandard |            20 | 808.0 ms | 3.01 ms | 2.82 ms |  1.00 |
| SortFullyInlined |            20 | 593.6 ms | 2.23 ms | 2.08 ms |  0.73 |
| SortSplitInlined |            20 | 643.8 ms | 3.05 ms | 2.85 ms |  0.80 |
|                  |               |          |         |         |       |
|     SortStandard |            30 | 748.6 ms | 2.88 ms | 2.41 ms |  1.00 |
| SortFullyInlined |            30 | 565.9 ms | 2.54 ms | 2.38 ms |  0.76 |
| SortSplitInlined |            30 | 617.4 ms | 3.49 ms | 3.26 ms |  0.82 |
|                  |               |          |         |         |       |
|     SortStandard |            40 | 726.1 ms | 5.10 ms | 4.77 ms |  1.00 |
| SortFullyInlined |            40 | 504.0 ms | 1.86 ms | 1.74 ms |  0.69 |
| SortSplitInlined |            40 | 576.4 ms | 2.98 ms | 2.64 ms |  0.79 |
|                  |               |          |         |         |       |
|     SortStandard |            50 | 671.1 ms | 6.76 ms | 6.32 ms |  1.00 |
| SortFullyInlined |            50 | 456.7 ms | 1.51 ms | 1.34 ms |  0.68 |
| SortSplitInlined |            50 | 548.9 ms | 2.18 ms | 1.82 ms |  0.82 |

I'm not sure that having 50% NaNs (or even 10%!) is a realistic scenario, but it does illustrate some divergence in results between fully- and split-inlined.

I didn't go above 50% NaNs because by the time you reach 100% NaNs the data is fully sorted and then what's the point of sorting, and it didn't seem like a realistic scenario either.

**Comment by rickbrew** [comment]
So this illustrates that _for sorting an array of elements_, inlining `CompareTo()` has great results. It's hard to say what the impact would be in other scenarios that have more/less reliance on `CompareTo()` for their performance.

**Comment by tannergooding** [comment]
Given the above, and that split inlining is still a lot faster than no inlining, I think its reasonable as a stop gap.

@danmoseley, do you have any concerns with us merging this for RC1? If the official benchmarks show any regression not covered here or on older hardware, we'd likely need to pull it back out.

File: src/libraries/System.Private.CoreLib/src/System/Double.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: Does the manual split inlining actually make the code significantly smaller in real world situations vs. just aggressive inlining the whole thing? The call introduced by the split inlining is going to have secondary effect like extra register spilling, so it is not obvious to me that it is actually profitable.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: The entire inlined assembly is fairly large: https://github.com/dotnet/runtime/pull/56501#issuecomment-890001797

`NaN` checks in particular aren't done very efficiently today and we should try to improve it in .NET 7 so the call will remove 2-4 extra branches from the inlined code-path (in favor of a call which may spill on a probably unlikely path).

In the link, I gave an example of what we could generate for this entire comparison if we did something "more optimally".

Comment (rickbrew) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: @jkotas In sharplab, split inlining results in smaller IL but larger machine code https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRIpgDMFbGXZkA3kTJ6KqzABYyAWQAUASm279AXyIPCS1SgRC6UhGYCeAEU46YEEiHUJ9Mh4hKHEIBBYyAKDBMjAAfVj+OhYFcP1bPQBtUxYEAAtOABMASV5+cxLyqtqefgB5HiZOGBQqADlOaph+WFgAc0sAXQKDMlhvLl4IIRYAFU5zJOCEzOzLGbCIiKgAMzJzdN2EgB4yK+tMAHYyAFpcXKP9U/PLiCyEgB8dz+ewoz3eM2OZwuGRBCQAvPDgf8Hs8SB9PmQAPRYsgAQW8glQ3m6CU4Z3KOzhKDmNL6ED6VEhX2hlUC2yo1RQ9L6MPu+zymP0TzIbOSbC5PPM9zIAH4yOQQK8IYKjix+CgWMzPiKVREnBEZsVShUanUGibmnUOl0elQ8WMxisUCgYiwhiMYOMpjMVHNGBpuDxlmtOAAxOj8fg+D2wFiVTbslL3A7aubQ37/Mi3GUit4Yz7fPlwshA3PggtHIuZ7JkRHI0Ei9FpnH4wksYlkUldillKn/GlQOkMpmqqHnMUcyUM4sogVCiIiyeCTncmcy+WK5WViLqzVpxcVmYG/Kq41NM2tC0XlrtTpQbq9B1O2Su8Tu4ajGATaaqv3zQMlhWdYRFaKAEFjGB43MZd+z2VMxxZH5YSzHM4VRbc02rFDazLdCwTIPVMWwq46yRcsFR3YVnkWYNgM4AB1cCygJAAZDs3DaKCpXuSsnBmKIYjiBIANokN1iY8o2I4hAuJYKVYIbFh530Q5MVbAkyCJNwuygnsyEpJTB2HRksNZJMJTXXka2UlSFwI2DVx4/DNzIJV8zTPctUQnUj1VJw7CAA===

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: I don't have a particular preference on which we go with or if we hold off and wait for the JIT to make this more intrinsic or even simply improve the general handling for these kinds of checks and branches.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdxFAwCyACgCUx0+YC+Nd9S27cGKBzAMBgAVVQwaE2pzJnIkBmAICEkGAElcWWgMGAATAH0RcWxYYIgAMQ5JSQBPFK5JXi4cuzYYbGyAeTqq2TFsLgAebIgOYGkAPgZsBxczSOjogDNoBjsGoN4GAF4GOjUGDf7JlgAZGC4AcwwACwY4BnI93gBqJ+mo+Y/eBZWC3uKyirVWr1RrZOzYADavAAumhJlCGE97tCnIc6E5iAB2BgLbCSXAwDTvD5Yhh+DiEmYMVzRKlUiE2GDXCDZFLiSR2RnM1ns9piPgQLi4FgAQXO51guFwvAUMGBDQa5wc0KpOhicTWBlEfxgJXKlRqdQaTQaDAAIsNRjAGPhcgo8RS4RaRtIGPbJBS3tE5h99t87Lb3RSGIcgzAMdi4A8qdEvitAw7rRMwxH7kTfX743bE1ttimmNjdjHzAB6EsMEVBaR4IKC60Qb7Xa1h3D7VsAOWw7ZYxbMcbsQxdgjSnfbAezHvDXoz5lJg6tLBHXbs+YA/DsGCBbtHifMYPiYL35qSd9FPLTd8w4gkkql0pkcvltUVdRAevUMPLQc1Wh0uj0+kGS1xkmadZiPJYoBWTUNm2XZ9hDI5Tgua5bjTfYXjAmd+1+F8SnfXhPyNb9IRhOFSMRZFUR2VNcQPdNfVJclKV3GlzE0XcGSZK4WTZMQOS5HieX4vkBSFUVxUlaVZS/RVlVVSxYn2HgtUKf4CKIkETS4c1gOtBNJzhedXRTKkfU+f0DODUNE1TKMGIsrMwwYZNbILNMj37KzrU2PM3NJItd2iUlcP+AB1QirkrU4a06GBR3HMM4RTBzqQ46IxCgGVsCydVlKCULXwi65otaXw4oS01nStG0J0dBhjObWyzKPMsKyrMra0aBgGzJK4msnVteA7LseyC8x+0axdcAS7yHCwjM5z06aErXDct3so99wJI9guxU83A8IA=

shows the behavior of the two options in a simple real world case. Notice that the `IsSorted_CompareToFullyInlined` (code size 0x59) produces smaller code than `IsSorted_CompareToSplitInlined` (code size 0x6f). The obvious problem is that the slow path is looks like a good inlining candidate and the JIT decides to inline it. Once you try to fix it by marking the slow path with NoInlining:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdxFAwCyACgCUx0+YC+Nd9S27cGKBzAMBgAVVQwaE2pzJnIkBmAICEkGAElcWWgMGAATAH0RcWxYYIgAMQ5JSQBPFK5JXi4cuzYYbGyAeTqq2TFsLgAebIgOYGkAPgZsBxczSOjogDNoBjsGoN4GAF4GOjUGDf7JlgAZGC4AcwwACwY4BnI93gBqJ+mo+Y/eBZWC3uKyirVWr1RrZOzYADavAAumhJlCGE97tCnIc6E5iAB2BgLbCSXAwDTvD5Yhh+DiEmYMVzRKlUiE2GDXCDZFLiSR2RnM1ns9piPgQLi4FgAQXO51guFwvAUMGBDQa5wc0KpOhicTWBlEfxgJXKlRqdQaTQaDAAIsNRjAGPhcgo8RS4RaRtIGPbJBS3tE5h99t87Lb3RSGIcgzAMdi4A8qdEvitAw7rRMwxH7kTfX743bE1ttimmNjdjHzAB6EsMEVBaR4IKC60Qb7Xa1h3D7VsAOWw7ZYxbMcbsQxdgjSnfbAezHvDXoz5lJg6tLBHXbs+YA/DsGCBbtHifMYPiYL35qSd9FPLTd8w4gkkql0pkcvltUVdRAevUMPLQc1Wh0uj0+kGS1xkmadZiPJYoBWTUNm2XZ9hDI5Tgua5bjTfYXjAmd+1+F8SnfXhPyNb9IRhOFSMRZFUR2VNcQPdNfVJclKV3GlzE0XcGSZK4WTZMQOS5HieX4vkBSFUVxUlaVZS/RVlVVSxYn2HgtUKf4CKIkETS4c1gOtBNJzhedXRTKkfU+f0DODUNE1TKMGIsrMwwYZNbILNMj37KzrU2PM3NJItd2iUlcP+AB1QirkrU4a06GBR3HMM4RTBzqQ46IuO5PiBO43jeX5XhBWFdsIFki55N3MQoBlbAsnVZSglC18IuuaLWl8OKEtNZ0rRtCdHQYYzm1ssyjzLCsq3a2tGgYBsySuYbJ1bXgOy7HsgvMfshsXXAEu8hwsIzOc9J2hK1w3Ld7KPfcCSPYLsVPNwPCAA

it gets a bit better, but the code size of the non-split inlined version is still smaller.

> I don't have a particular preference on which we go with or if we hold off and wait for the JIT to make this more intrinsic

I agree that it would be nice to teach the JIT to deal with the efficiently. If we want to tweak the performance by manual inlining, simple AggressiveInlining is better than the manual split inlinining as my examples demonstrated.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: > If we want to tweak the performance by manual inlining, simple AggressiveInlining should be better than the manual split inlinining as my examples demonstrated.

I'm fine with this. I would like to disagree on the examples adequately demonstrating the difference however as smaller code isn't necessarily better.

There are many factors that can impact perf here including the likelihood that a given path will be hit (`NaNs` are likely rare) and considerations such as the number of branches in a "small window" (16-32 aligned bytes of assembly). This was why I was initially hesitant about the change as it could regress certain "hot loops" due to the additional branches (`7`) that would now be present directly in the loop. Of course, the partial inlining could also impact this in interesting ways with its overall larger codegen but it does reduce the branch count by `3`.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: I agree that you would be likely able to come up with cases where the split inlining is better due for micro-architecture reasons, if you tried hard enough. The data we have so far in this thread is that a simple AggressiveInlining produces faster and smaller code.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs:200]: To satisfy my own curiosity, I tried it in double.cs in corelib.  Here main is what's currently checked in, pr1 is CompareTo aggressively inlined, and pr2 is it split.

```C#
private double[] _doubles = Enumerable.Range(0, 1_000_000).Select(i => i * 1.23).ToArray();
private double[] _scratch = new double[1_000_000];

[Benchmark]
public bool IsSorted()
{
    ReadOnlySpan<double> a = _doubles;
    for (int i = 0; i < a.Length - 1; i++)
        if (a[i].CompareTo(a[i + 1]) > 0)
            return false;
    return true;
}

[Benchmark]
public void CopyAndSort()
{
    _doubles.CopyTo(_scratch, 0);
    Array.Sort(_scratch);
}

[Benchmark]
public int Search() => Array.BinarySearch(_doubles, 2_000_000);

[Benchmark]
public int CompareSequence() => _doubles.AsSpan().SequenceCompareTo(_doubles);
```

|          Method |        Toolchain |             Mean | Ratio | Code Size |
|---------------- |----------------- |-----------------:|------:|----------:|
|        IsSorted | main\corerun.exe |  1,921,366.71 ns |  1.00 |     155 B |
|        IsSorted |  pr1\corerun.exe |    739,726.37 ns |  0.39 |      97 B |
|        IsSorted |  pr2\corerun.exe |    970,601.44 ns |  0.50 |     117 B |
|                 |                  |                  |       |           |
|     CopyAndSort | main\corerun.exe | 11,761,377.29 ns |  1.00 |   1,060 B |
|     CopyAndSort |  pr1\corerun.exe | 11,810,071.67 ns |  1.00 |   1,060 B |
|     CopyAndSort |  pr2\corerun.exe | 11,837,290.42 ns |  1.01 |   1,060 B |
|                 |                  |                  |       |           |
|          Search | main\corerun.exe |         58.53 ns |  1.00 |     207 B |
|          Search |  pr1\corerun.exe |         35.51 ns |  0.61 |     207 B |
|          Search |  pr2\corerun.exe |         38.65 ns |  0.66 |     207 B |
|                 |                  |                  |       |           |
| CompareSequence | main\corerun.exe |  1,829,633.22 ns |  1.00 |     330 B |
| CompareSequence |  pr1\corerun.exe |  1,452,953.83 ns |  0.79 |     306 B |
| CompareSequence |  pr2\corerun.exe |  1,523,349.48 ns |  0.83 |     316 B |

I asked whether it would make sense to split it.  I'm fine with the answer being "no" :smile:

Comment (rickbrew) [reply src/libraries/System.Private.CoreLib/src/System/Double.cs]: Alright I have reverted the commit that split the inlining, so we're back to just regular aggressive inlined

**Review by jkotas — APPROVED** [review]

**Comment by danmoseley** [comment]
> @danmoseley, do you have any concerns with us merging this for RC1? If the official benchmarks show any regression not covered here or on older hardware, we'd likely need to pull it back out

Nope, it'd very low risk if we have to pull it out. If you've concluded it's a worthwhile change in the real world which it sounds like you have.

File: src/libraries/System.Private.CoreLib/src/System/Single.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Single.cs:186]: @tannergooding, @pgovind, should we be doing the same for Half.CompareTo?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Single.cs:186]: I don't think its as importantas `Half.CompareTo` is quite a bit more expensive. Each operator is an actual call, not a primitive operation that compiles down to a CPU instruction

**Review by tannergooding — APPROVED** [review]

**Comment by stephentoub** [comment]
@tannergooding, is the plan to backport this?

**Comment by jeffhandley** [comment]
@stephentoub Yep. @tannergooding and I chatted about it earlier. Before we merge into `release/6.0` though, I'd like to make sure the latest perf run against `main` doesn't indicate any unintended regressions (even though likelihood is low).

**Comment by GrabYourPitchforks** [comment]
> Before we merge into `release/6.0` though, I'd like to make sure the latest perf run against `main` doesn't indicate any unintended regressions

As @rickbrew mentioned earlier, we may see a minor size-on-disk regression as would accompany most "this method is now inlined where previously it wasn't" changes. I assume we're still ok with that?

**Comment by jeffhandley** [comment]
Assuming it's truly _minor_, yeah.

**Pull request merged**
